### PR TITLE
Materials for Range tutorial

### PR DESCRIPTION
**Pull request by gahjelle**
**Where to put new files:**

- New files should go into a top-level subfolder, named after the article slug. For example: `my-awesome-article`

**How to merge your changes:** 

1. [Make sure the CI code style tests all pass (+ run the automatic code formatter if necessary).](https://github.com/realpython/materials/blob/master/README.md)
2. Find an RP Team member on Slack and ask them to review & approve your PR.
3. Once the PR has one positive ("approved") review, GitHub lets you merge the PR.
4. 🎉

### CODE REVIEW

File: python-range/float_range.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
+import math
+from dataclasses import dataclass, field
+
+
+@dataclass
+class FloatRange:
+    start: float | int
+    stop: float | int = None
```
Comment (bzaczynski) [inline python-range/float_range.py:8]: This is technically `float | int | None`.

File: python-range/float_range.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        return round((element - self.start) / self.step)
+
+
+@dataclass
+class _FloatRangeIterator:
+    start: int
+    stop: int
+    step: int
+    _num_steps: int = field(default=0, init=False)
+
```
Comment (bzaczynski) [inline python-range/float_range.py:87]: The iterator protocol requires that an iterator object implements the `.__iter__()` method as well:

```suggestion
    def __iter__(self):
        return self
```

Without it, you won't be able to loop over the iterator itself or pass it into a `list()`:

```python
>>> for value in _FloatRangeIterator(1, 3, 0.5):
...     print(value)
...
1.0
1.5
2.0
2.5

>>> it = _FloatRangeIterator(1, 3, 0.5)
>>> list(it)
[1.0, 1.5, 2.0, 2.5]
```

File: python-range/float_range.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    def __next__(self):
+        """Calculate the next element in the iteration."""
+        element = self.start + self._num_steps * self.step
+        if any(
+            self.step > 0 and element >= self.stop,
+            self.step < 0 and element <= self.stop,
```
Comment (bzaczynski) [inline python-range/float_range.py:93]: When `.step` is equal to zero, the iteration will never stop. The original `range()` function rejects zero as the step, so you could perhaps add similar validation in the `.__post_init__()` method.

File: python-range/float_range.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    stop: float | int = None
+    step: float | int = 1.0
+
+    def __post_init__(self):
+        """Validate parameters."""
+        # Only one argument is given
+        if self.stop is None:
+            self.stop = self.start
+            self.start = 0
+
```
Comment (bzaczynski) [inline python-range/float_range.py:17]: If step is zero, we should raise an error to handle edge cases:

```python
if math.isclose(self.step, 0):
    raise ValueError("'step' must not be zero")
```

**Review by bzaczynski — CHANGES_REQUESTED** [review]
@gahjelle Overall, your code looks good to me. I only flagged a few minor edge cases that it would be nice to handle. Let me know what you think. Thanks!

File: python-range/float_range.py (hunk 4 above)
Comment (gahjelle) [reply python-range/float_range.py:17]: This is checked by `if not isinstance(self.step, float | int) or self.step == 0` below. I'm switching to `isclose()`.

File: python-range/float_range.py (hunk 3 above)
Comment (gahjelle) [reply python-range/float_range.py:93]: I've emphasized in the docstring that `_FloatRangeIterator` is non-public and should only be instantiated through `FloatRange` which does the validation. I'll leave the foot-gun for those desperate to use it ...

**Comment by gahjelle** [comment]
Thanks @bzaczynski for the detailed review. I've updated the code.

**Review by bzaczynski — APPROVED** [review]
@gahjelle Thank you for the updates. They looks good to me :+1:

**Pull request merged**